Update for pyarrow 21.0.0

AlenkaF opened the pull request:
In PyArrow version 21.0.0 `PyExtensionType` will be removed, see https://github.com/apache/arrow/pull/46199. We need to update the checks in `__init__` so that the hotfix doesn't break with future pyarrow versions.

AlenkaF committed:
Check for PyExtensionType

pitrou commented:
Hi @AlenkaF , two things:
1) can you merge from main to fix existing CI tests (I had to [update them](https://github.com/pitrou/pyarrow-hotfix/commit/7f855ee43accd1cf0a0466a7df0891c94f673102) to account for GitHub Actions runner updates)
2) can you check the hotfix tests run with PyArrow 21?

AlenkaF committed:
Update tests

AlenkaF commented:
> can you merge from main to fix existing CI tests (I had to [update them](https://github.com/pitrou/pyarrow-hotfix/commit/7f855ee43accd1cf0a0466a7df0891c94f673102) to account for GitHub Actions runner updates)

Done.

> can you check the hotfix tests run with PyArrow 21?

With the [last commit](https://github.com/pitrou/pyarrow-hotfix/pull/8/commits/28b8b254df2e4abe60168e90cc090bd31eb9e993) and hacky diff for local dev setup:

```
diff --git a/tests/test_in_process.py b/tests/test_in_process.py
index 19bf894..0f87759 100644
--- a/tests/test_in_process.py
+++ b/tests/test_in_process.py
@@ -9,7 +9,7 @@ import pyarrow.parquet as pq
 import pyarrow_hotfix
 
 
-pa_version = tuple(map(int, pa.__version__.split('.')))
+pa_version = tuple(map(int, ("21.0.0").split('.')))
```

the tests passed. 

```
> pytest tests
========================================================================== test session starts ==========================================================================
platform darwin -- Python 3.12.10, pytest-8.3.5, pluggy-1.5.0
rootdir: /Users/alenkafrim/Repos/pyarrow-hotfix
configfile: pyproject.toml
plugins: anyio-4.9.0, hypothesis-6.127.4
collected 14 items                                                                                                                                                      

tests/test_in_process.py .......                                                                                                                                  [ 50%]
tests/test_subprocess.py .......                                                                                                                                  [100%]

========================================================================== 14 passed in 2.00s ===========================================================================
```

Adding package info, just in case:

```
>>> pip list
...
pyarrow                       20.0.0.dev327+gb4b3728ff4 /Users/alenkafrim/Repos/arrow/python
pyarrow-hotfix                0.6                       /Users/alenkafrim/Repos/pyarrow-hotfix
```

pitrou commented:
It seems this needs to be fixed for PyArrow 0.14, see CI failures.

AlenkaF commented:
Plan to have a look at it tomorrow/Friday. Thanks!

AlenkaF committed:
Move the 21.0.0 check under unregister_extension_type check

AlenkaF commented:
@pitrou pinging for CI approval.

raulcd reviewed (approved):
Thanks @AlenkaF !

pitrou commented:
I'm gonna make a new release soon.

pitrou commented:
The release is now made: https://pypi.org/project/pyarrow-hotfix/
